infrastructure/comms/src/peer_manager/node_id.rs

```diff
@@ -0,0 +1,243 @@
+//  Copyright 2019 The Tari Project
+//
+//  Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+//  following conditions are met:
+//
+//  1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+//  disclaimer.
+//
+//  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use derive_error::Error;
+use digest::Digest;
+use tari_crypto::keys::PublicKey;
+
+#[derive(Debug, Error)]
+pub enum NodeIdError {
+    OutOfBounds,
+}
+
+#[derive(Clone, Debug, Eq, PartialOrd, Ord)]
+pub struct NodeDistance([u8; 32]);
+
+impl NodeDistance {
+    /// Construct a new zero distance
+    pub fn new() -> NodeDistance {
+        (NodeDistance::from(&[0; 32]))
+    }
+
+    /// Construct a node distance from an array of 32 bytes
+    pub fn from(array: &[u8; 32]) -> NodeDistance {
+        (NodeDistance(*array))
+    }
+
+    /// Calculate the distance between two node ids using the XOR metric
+    pub fn from_node_ids(x: &NodeId, y: &NodeId) -> NodeDistance {
+        let mut nd = NodeDistance::new();
+        for i in 0..nd.0.len() {
+            nd.0[i] = x.0[i] ^ y.0[i];
+        }
+        (nd)
+    }
+}
+
+impl PartialEq for NodeDistance {
+    fn eq(&self, nd: &NodeDistance) -> bool {
+        (self.0 == nd.0)
+    }
+}
+
+#[derive(Clone, Debug, Eq)]
+pub struct NodeId([u8; 32]);
+
+impl NodeId {
+    /// Construct a new node id on the origin
+    pub fn new() -> NodeId {
+        (NodeId::from(&[0; 32]))
```

Comment:
Suggestion: NodeId and NodeDistance need to have the same length, so if you Alias the array to a type, you only have to change it in one place if we ever decide to change the length of the array

Reply:
Made use of shared type.